test/infrastructure/vcsim/controllers/virtualmachine_controller.go

```diff
@@ -0,0 +1,326 @@
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the owner VSphereMachine.
+	vSphereMachine, err := util.GetOwnerVMWareMachine(ctx, r.Client, virtualMachine.ObjectMeta)
+	// vsphereMachine can be nil in cases where custom mover other than clusterctl
+	// moves the resources without ownerreferences set
+	// in that case nil vsphereMachine can cause panic and CrashLoopBackOff the pod
+	// preventing vspheremachine_controller from setting the ownerref
+	if err != nil || vSphereMachine == nil {
+		log.Info("Owner VSphereMachine not found, won't reconcile", "key", req.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+	log = log.WithValues("VSphereMachine", klog.KObj(vSphereMachine))
+
+	// Fetch the Machine.
+	machine, err := capiutil.GetOwnerMachine(ctx, r.Client, vSphereMachine.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if machine == nil {
+		log.Info("Waiting for Machine Controller to set OwnerRef on VSphereMachine")
+		return ctrl.Result{}, nil
+	}
+	log = log.WithValues("Machine", klog.KObj(machine))
+
+	// Fetch the Cluster.
+	cluster, err := capiutil.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		log.Info("VSphereMachine owner Machine is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info(fmt.Sprintf("Please associate this machine with a cluster using the label %s: <name of cluster>", clusterv1.ClusterNameLabel))
+		return ctrl.Result{}, nil
+	}
+	log = log.WithValues("Cluster", klog.KObj(cluster))
+
+	// Return early if the object or Cluster is paused.
+	if annotations.IsPaused(cluster, vSphereMachine) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Fetch the VSphereCluster.
+	key := client.ObjectKey{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+	vsphereCluster := &vmwarev1.VSphereCluster{}
+	if err := r.Client.Get(ctx, key, vsphereCluster); err != nil {
+		log.Info("VSphereCluster can't be retrieved")
+		return ctrl.Result{}, err
+	}
+	log = log.WithValues("VSphereCluster", klog.KObj(vsphereCluster))
+
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Compute the resource group unique name.
+	// NOTE: We are using reconcilerGroup also as a name for the listener for sake of simplicity.
+	resourceGroup := klog.KObj(cluster).String()
+
+	// Check if there is a conditionsTracker in the resource group.
+	// The conditionsTracker is an object stored in memory with the scope of storing conditions used for keeping
+	// track of the provisioning process of the fake node, etcd, api server, etc for this specific virtualMachine.
+	// (the process managed by this controller).
+	inmemoryClient := r.InMemoryManager.GetResourceGroup(resourceGroup).GetClient()
+	conditionsTracker := &infrav1.VSphereVM{} // NOTE: The type of this object doesn't matter as soon as it implements Cluster API's conditions interfaces.
```

Comment:
Wondering if using something else would be less confusing :) (supervisor vs. infrav1.VSphereVM)

E.g. a vmoprv1.VirtualMachine{}

(or maybe define a separate struct)

Reply:
improved the comment
unfortunately vmoprv1.VirtualMachine isn't a condition getter